### PR TITLE
AMD ROCm: Bump rocm-tensile with 3.9.0, 3.10.0, 4.0.0 releases

### PR DESCRIPTION
**Pull request by arjun-raj-kuppala**

### CODE REVIEW

File: var/spack/repos/builtin/packages/rocm-tensile/package.py [hunk 1]
```diff
@@ -33,29 +36,48 @@ class RocmTensile(CMakePackage):
         depends_on('rocm-smi@' + ver, type='build', when='@' + ver)
         depends_on('rocm-smi-lib@' + ver, type='build', when='@' + ver)
         depends_on('llvm-amdgpu@' + ver, type='build', when='@' + ver)
-
     # This is the default library format since 3.7.0
     depends_on('msgpack-c@3:', when='@3.7:')
     depends_on('boost@1.58.0', type=('build', 'link'))
+    depends_on('llvm-openmp', type=('build', 'link'))
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rocm-tensile/package.py]: this is not the way to do it, the compiler should provide the omp libs if they are strictly necessary, they are currently not link dependencies like here.

**Review by haampie — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/rocm-tensile/002-fix-boost-inc-dirs.patch [hunk 2]
```diff
@@ -0,0 +1,13 @@
+diff --git a/Tensile/Source/client/CMakeLists.txt b/Tensile/Source/client/CMakeLists.txt
+index 9ccba86..59ed2e2 100644
+--- a/Tensile/Source/client/CMakeLists.txt
++++ b/Tensile/Source/client/CMakeLists.txt
+@@ -32,7 +32,7 @@ if(NOT ROCmSMI_FOUND)
+     find_package(ROCmSMI REQUIRED)
+ endif()
+ 
+-target_include_directories(TensileClient PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/include")# "${rocm_smi_root}/include")
++target_include_directories(TensileClient PUBLIC "${Boost_INCLUDE_DIRS}" "${CMAKE_CURRENT_SOURCE_DIR}/include")# "${rocm_smi_root}/include")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rocm-tensile/002-fix-boost-inc-dirs.patch]: This is not necessary, it uses `find_package(Boost  COMPONENTS program_options filesystem REQUIRED)`, should not require a patch.

Comment (arjun-raj-kuppala) [reply var/spack/repos/builtin/packages/rocm-tensile/002-fix-boost-inc-dirs.patch]: Thanks for the review. I had removed this patch

File: var/spack/repos/builtin/packages/rocm-tensile/package.py (hunk 1 above)
Comment (arjun-raj-kuppala) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: I will remove the link dependency with a update.

File: var/spack/repos/builtin/packages/rocm-tensile/package.py [hunk 3]
```diff
@@ -33,29 +36,48 @@ class RocmTensile(CMakePackage):
...
         args = [
             '-Damd_comgr_DIR={0}'.format(self.spec['comgr'].prefix),
             '-DTensile_COMPILER=hipcc',
-            '-DTensile_ARCHITECTURE={0}'.format(archs),
+            '-DTensile_ARCHITECTURE={0}'.format(arch),
             '-DTensile_LOGIC=asm_full',
             '-DTensile_CODE_OBJECT_VERSION=V3',
-            '-DBoost_USE_STATIC_LIBS=Off',
+            '-DBoost_USE_STATIC_LIBS=off',
+            '-DCMAKE_CXX_FLAGS:String=-I{0}'.format(boost_incl),
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rocm-tensile/package.py]: Boost should really be found by cmake, and we should avoid adding compiler flags like this. What is the problem here you're trying to solve here?

Comment (arjun-raj-kuppala) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: Iam running into below issue if i dont add  **'-DCMAKE_CXX_FLAGS:String=-I{0}'.format(boost_incl),**


```
cd /tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-build-xoe7npt/client && /tmp/spack/opt/spack/linux-sles15-zen/gcc-7.5.0/hip-3.10.0-wysrhl5hb3y6345raqhdj3mg5jyaal7m/bin/hipcc -DTENSILE_DEFAULT_SERIALIZATION -DTENSILE_STATIC_ONLY -DTENSILE_USE_HIP -DTENSILE_YAML=1 -D__HIP_PLATFORM_HCC__=1 -D__HIP_ROCclr__=1 -I/tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/include -I/tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/lib/include -isystem /tmp/spack/opt/spack/linux-sles15-zen/gcc-7.5.0/rocm-smi-lib-3.10.0-dn6j5fjh2imo6myb6qh54vmv4c4wdkmx/include -fopenmp -std=c++14 -O2 -g -DNDEBUG -mno-xnack -Xarch_gfx906 -msram-ecc -Xarch_gfx908 -msram-ecc -xhip --cuda-gpu-arch=gfx900 --cuda-gpu-arch=gfx906 --cuda-gpu-arch=gfx908 -o CMakeFiles/TensileClient.dir/source/TimingEvents.cpp.o -c /tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/source/TimingEvents.cpp
In file included from /tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/source/DataInitialization.cpp:27:
/tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/include/DataInitialization.hpp:29:10: fatal error: 'boost/program_options.hpp' file not found
#include <boost/program_options.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/source/CSVStackFile.cpp:27:
/tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/include/CSVStackFile.hpp:34:10: fatal error: 'boost/lexical_cast.hpp' file not found
#include <boost/lexical_cast.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from /tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/source/ResultFileReporter.cpp:27:
In file included from /tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/include/ResultFileReporter.hpp:29:
/tmp/root/spack-stage/spack-stage-rocm-tensile-3.10.0-xoe7npteuwocnb6hwo6r6xqslutqn4en/spack-src/Tensile/Source/client/include/CSVStackFile.hpp:34:10: fatal error: 'boost/lexical_cast.hpp' file not found
#include <boost/lexical_cast.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated when compiling for gfx900.
make[2]: *** [client/CMakeFiles/TensileClient.dir/build.make:98: client/CMakeFiles/TensileClient.dir/source/CSVStackFile.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
1 error generated when compiling for gfx900.
make[2]: *** [client/CMakeFiles/TensileClient.dir/build.make:137: client/CMakeFiles/TensileClient.dir/source/DataInitialization.cpp.o] Error 1
1 error generated when compiling for gfx900.

```

Comment (alalazo) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: Which version of `boost` are you using? Is the issue reproducible with other specs built with CMake against boost at the same version?

Comment (haampie) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: This is not the way to fix that problem. Please figure out why https://github.com/ROCmSoftwarePlatform/Tensile/blob/257ff1f98d8cb0f6361b40f5313bdf9f94f4155b/Tensile/Source/client/CMakeLists.txt#L37 is not adding the include flags to those targets.

**Comment by haampie** [comment]
@arjun-raj-kuppala: with regard to openmp, what you want to do is use openmp as provided by the compiler. In your case the compiler is from the llvm-amdgpu package as you're using hipcc, right? If so, can't you build llvm-amdgpu with openmp support? If you take a look at the llvm package in spack, what they do is 
```
        cmake_args.append(
            "-DLLVM_ENABLE_PROJECTS:STRING={0}".format(";".join(projects))
        )
```
where projects includes:

```
        if "+clang" in spec:
            projects.append("clang")
            projects.append("clang-tools-extra")
            projects.append("openmp")
```
See https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/llvm/package.py#L438

Can't you do the same for llvm-amdgpu?

**Comment by haampie** [comment]
@arjun-raj-kuppala can you include https://github.com/ROCmSoftwarePlatform/Tensile/pull/1254 as a patch after adding openmp support to llvm-amdgpu? That would be the way forward here.

**Review by haampie — CHANGES_REQUESTED** [review]
See above for requested changes.

File: var/spack/repos/builtin/packages/rocm-tensile/package.py (hunk 3 above)
Comment (arjun-raj-kuppala) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py:78]: @haampie Sure will check and create a separate PR for llvm-amdgpu recipe update and also add above patch. In the meantime, can we merge this in.

**Comment by arjun-raj-kuppala** [comment]
> @arjun-raj-kuppala can you include [ROCmSoftwarePlatform/Tensile#1254](https://github.com/ROCmSoftwarePlatform/Tensile/pull/1254) as a patch after adding openmp support to llvm-amdgpu? That would be the way forward here.

I have implemented the request changes. @haampie  Can you review and approve this ?

**Comment by arjun-raj-kuppala** [comment]
@haampie @adamjstewart 

Can any one of you help approve this. This is pending for some time now. Let me know if anything needs to be done here.

File: var/spack/repos/builtin/packages/rocm-tensile/package.py [hunk 4]
```diff
@@ -14,48 +14,54 @@ class RocmTensile(CMakePackage):
...
     version('3.5.0', sha256='71eb3eed6625b08a4cedb539dd9b596e3d4cc82a1a8063d37d94c0765b6f8257')
 
-    amdgpu_targets = ('all', 'gfx803', 'gfx900', 'gfx906', 'gfx908')
+    tensile_architecture = ('all', 'gfx803', 'gfx900', 'gfx906', 'gfx908')
 
-    variant('amdgpu_target', default='all', multi=True, values=amdgpu_targets)
+    variant('tensile_architecture', default='all', values=tensile_architecture, multi=False)
 
     depends_on('cmake@3:', type='build')
+    depends_on('boost@1.58.0', type=('build', 'run'), when='@3.9.0:')
```
Comment (haampie) [inline var/spack/repos/builtin/packages/rocm-tensile/package.py]: It's already listed a few lines below
```suggestion

```
also I don't understand `type=run` for boost.

File: var/spack/repos/builtin/packages/rocm-tensile/package.py (hunk 3 above)
Comment (arjun-raj-kuppala) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: I agree this should be looked at and requested at source. But this is a tagged release so even if get them to add the include directory there would require a patch for the current releases. Is my understanding correct? If yes, should we be able to merge this ?

File: var/spack/repos/builtin/packages/rocm-tensile/package.py (hunk 4 above)
Comment (arjun-raj-kuppala) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: Let me check this..

**Comment by haampie** [comment]
Hi @arjun-raj-kuppala, can you drop fixing boost to a specific version? The latest boost seems to work just fine.

Also, can you remove setting those include flags for Boost? CMake should be able to locate it.

Lastly, I'm thinking you should be careful with OpenMP. I don't know what it's being used for, but because we're setting CXX to hipcc, that means that if you link it to other libraries that use OpenMP too with the standard host compiler (say gcc), you run into problems. So maybe make just add a `+openmp` variant so users can turn it off?

**Comment by arjun-raj-kuppala** [comment]
> Hi @arjun-raj-kuppala, can you drop fixing boost to a specific version? The latest boost seems to work just fine.
> 
> Also, can you remove setting those include flags for Boost? CMake should be able to locate it.

Sure, will try without fixing the boost to specific version. But without including the include flag, i was getting error which i mentioned above.

**Comment by haampie** [comment]
Hm ok, I have succesfully built it without those flags on ubuntu 20.04

**Comment by arjun-raj-kuppala** [comment]
> Hm ok, I have succesfully built it without those flags on ubuntu 20.04

This is not working for me. I am running this on SLES 15

File: var/spack/repos/builtin/packages/rocm-tensile/package.py (hunk 4 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/rocm-tensile/package.py]: Can you still remove this line? Boost is added as dependency on line 42

**Comment by haampie** [comment]
Can you please figure out what's going wrong? Does it even detect Boost correctly?

```bash
$ spack install --keep-stage rocm-tensile
$ cd `spack location -s rocm-tensile`
$ grep -i boost */CMakeCache.txt
```
Thanks.

**Comment by arjun-raj-kuppala** [comment]
> Can you please figure out what's going wrong? Does it even detect Boost correctly?
> 
> ```shell
> $ spack install --keep-stage rocm-tensile
> $ cd `spack location -s rocm-tensile`
> $ grep -i boost */CMakeCache.txt
> ```
> 
> Thanks.

Issue was with using this version of boost@1.58.0. Once i changed it default boost the issue doesn't show up. 
With this i kind off addressed all the requested changes. Can you review and approve.

**Review by haampie — APPROVED** [review]

**Pull request merged**